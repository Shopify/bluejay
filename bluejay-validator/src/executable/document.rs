mod error;
mod orchestrator;
mod path;
mod rule;
pub mod rules;
mod variable_definition_input_type;
mod visitor;

pub use error::{ArgumentError, DirectiveError, Error};
pub use orchestrator::{BuiltinRulesValidator, Orchestrator};
pub use path::{Path, PathRoot};
pub use rule::Rule;
pub use rules::BuiltinRules;
pub use variable_definition_input_type::VariableDefinitionInputType;
pub use visitor::Visitor;
