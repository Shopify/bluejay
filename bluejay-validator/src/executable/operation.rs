mod analyzer;
pub mod analyzers;
mod orchestrator;
mod variable_values;
mod visitor;

pub use analyzer::Analyzer;
pub use orchestrator::Orchestrator;
pub use variable_values::{OperationDefinitionValueEvaluationExt, VariableValues};
pub use visitor::Visitor;
