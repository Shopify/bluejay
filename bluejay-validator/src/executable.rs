mod error;
mod rule;
mod rules;
mod validator;
mod visitor;

pub use error::Error;
pub use rule::Rule;
use rules::Rules;
pub use validator::{RulesValidator, Validator};
use visitor::Visitor;
