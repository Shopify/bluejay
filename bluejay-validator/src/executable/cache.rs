use crate::executable::VariableDefinitionInputType;
use bluejay_core::definition::SchemaDefinition;
use bluejay_core::executable::{
    ExecutableDocument, FragmentDefinition, OperationDefinition, VariableDefinition,
};
use bluejay_core::AsIter;
use std::collections::HashMap;

pub struct Cache<'a, E: ExecutableDocument, S: SchemaDefinition> {
    variable_definition_input_type_references:
        HashMap<&'a E::VariableType, VariableDefinitionInputType<'a, S::BaseInputType>>,
    indexed_fragment_definitions: HashMap<&'a str, &'a E::FragmentDefinition>,
}

impl<'a, E: ExecutableDocument, S: SchemaDefinition> Cache<'a, E, S> {
    pub fn new(executable_document: &'a E, schema_definition: &'a S) -> Self {
        let variable_definition_input_type_references =
            HashMap::from_iter(executable_document.operation_definitions().iter().flat_map(
                |operation_definition: &'a E::OperationDefinition| {
                    let variable_definitions_iterator = operation_definition
                        .as_ref()
                        .variable_definitions()
                        .map(|variable_definitions: &'a E::VariableDefinitions| -> <E::VariableDefinitions as AsIter>::Iterator<'a> {
                            variable_definitions.iter()
                        });

                    variable_definitions_iterator
                        .into_iter()
                        .flatten()
                        .filter_map(|variable_definition| {
                            let type_reference = variable_definition.r#type();
                            VariableDefinitionInputType::try_from((
                                schema_definition,
                                type_reference,
                            ))
                            .ok()
                            .map(|vditr| (type_reference, vditr))
                        })
                },
            ));
        let indexed_fragment_definitions = HashMap::from_iter(
            executable_document
                .fragment_definitions()
                .as_ref()
                .iter()
                .map(|fragment_definition| (fragment_definition.name(), fragment_definition)),
        );
        Self {
            variable_definition_input_type_references,
            indexed_fragment_definitions,
        }
    }

    pub fn variable_definition_input_type_reference(
        &self,
        variable_type: &E::VariableType,
    ) -> Option<&VariableDefinitionInputType<'a, S::BaseInputType>> {
        self.variable_definition_input_type_references
            .get(variable_type)
    }

    pub fn fragment_definition(&self, name: &str) -> Option<&'a E::FragmentDefinition> {
        self.indexed_fragment_definitions.get(name).copied()
    }
}
