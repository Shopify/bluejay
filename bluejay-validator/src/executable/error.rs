use bluejay_core::definition::{SchemaDefinition, TypeDefinitionReferenceFromAbstract};
use bluejay_core::executable::{ExecutableDocument, OperationDefinitionFromExecutableDocument};
use bluejay_core::OperationType;
#[cfg(feature = "parser-integration")]
use bluejay_parser::{
    ast::executable::ExecutableDocument as ParserExecutableDocument,
    error::{Annotation, Error as ParserError},
    HasSpan,
};

pub enum Error<'a, E: ExecutableDocument, S: SchemaDefinition> {
    NonUniqueOperationNames {
        name: &'a str,
        operations: Vec<&'a E::ExplicitOperationDefinition>,
    },
    NotLoneAnonymousOperation {
        anonymous_operations: Vec<&'a OperationDefinitionFromExecutableDocument<E>>,
    },
    SubscriptionRootNotSingleField {
        operation: &'a OperationDefinitionFromExecutableDocument<E>,
    },
    FieldDoesNotExistOnType {
        field: &'a E::Field,
        r#type: &'a TypeDefinitionReferenceFromAbstract<S::TypeDefinitionReference>,
    },
    FieldSelectionsDoNotMerge {
        selection_set: &'a E::SelectionSet,
    },
    OperationTypeNotDefined {
        operation: &'a E::ExplicitOperationDefinition,
    },
    LeafFieldSelectionNotEmpty {
        selection_set: &'a E::SelectionSet,
        r#type: &'a S::OutputTypeReference,
    },
    NonLeafFieldSelectionEmpty {
        field: &'a E::Field,
        r#type: &'a S::OutputTypeReference,
    },
}

#[cfg(feature = "parser-integration")]
impl<'a, S: SchemaDefinition> From<Error<'a, ParserExecutableDocument<'a>, S>> for ParserError {
    fn from(value: Error<'a, ParserExecutableDocument<'a>, S>) -> Self {
        match value {
            Error::NonUniqueOperationNames { name, operations } => Self {
                message: format!("Multiple operation definitions named `{name}`"),
                primary_annotation: None,
                secondary_annotations: operations
                    .iter()
                    .filter_map(|operation| {
                        operation.name().map(|operation_name| Annotation {
                            message: format!("Operation definition with name `{name}`"),
                            span: operation_name.span(),
                        })
                    })
                    .collect(),
            },
            Error::NotLoneAnonymousOperation {
                anonymous_operations,
            } => Self {
                message: "Anonymous operation not lone operation in document".to_string(),
                primary_annotation: None,
                secondary_annotations: anonymous_operations
                    .iter()
                    .map(|operation| Annotation {
                        message: "Anonymous operation definition".to_string(),
                        span: operation.selection_set().span(),
                    })
                    .collect(),
            },
            Error::SubscriptionRootNotSingleField { operation } => Self {
                message: "Subscription root is not a single field".to_string(),
                primary_annotation: Some(Annotation {
                    message: "Selection set contains multiple fields".to_string(),
                    span: operation.selection_set().span(),
                }),
                secondary_annotations: Vec::new(),
            },
            Error::FieldDoesNotExistOnType { field, r#type } => Self {
                message: format!(
                    "Field `{}` does not exist on type `{}`",
                    field.name().as_ref(),
                    r#type.name()
                ),
                primary_annotation: Some(Annotation {
                    message: format!("Field does not exist on type `{}`", r#type.name()),
                    span: field.name().span(),
                }),
                secondary_annotations: Vec::new(),
            },
            Error::OperationTypeNotDefined { operation } => Self {
                message: format!(
                    "Schema does not define a {} root",
                    OperationType::from(operation.operation_type()),
                ),
                primary_annotation: Some(Annotation {
                    message: format!(
                        "Schema does not define a {} root",
                        OperationType::from(operation.operation_type()),
                    ),
                    span: operation.operation_type().span(),
                }),
                secondary_annotations: Vec::new(),
            },
            Error::FieldSelectionsDoNotMerge { selection_set } => Self {
                message: "Field selections do not merge".to_string(),
                primary_annotation: Some(Annotation {
                    message: "Field selections do not merge".to_string(),
                    span: selection_set.span(),
                }),
                secondary_annotations: Vec::new(),
            },
            Error::LeafFieldSelectionNotEmpty {
                selection_set,
                r#type,
            } => Self {
                message: format!(
                    "Selection on field of leaf type `{}` was not empty",
                    r#type.as_ref().display_name()
                ),
                primary_annotation: Some(Annotation {
                    message: "Selection set on field of leaf type must be empty".to_string(),
                    span: selection_set.span(),
                }),
                secondary_annotations: Vec::new(),
            },
            Error::NonLeafFieldSelectionEmpty { field, r#type } => Self {
                message: format!(
                    "No selection on field of non-leaf type `{}`",
                    r#type.as_ref().display_name()
                ),
                primary_annotation: Some(Annotation {
                    message: "Fields of non-leaf types must have a selection".to_string(),
                    span: field.name().span(),
                }),
                secondary_annotations: Vec::new(),
            },
        }
    }
}
