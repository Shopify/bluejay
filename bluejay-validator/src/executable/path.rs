use bluejay_core::executable::ExecutableDocument;
use std::cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd};
use std::hash::{Hash, Hasher};

pub struct Path<'a, E: ExecutableDocument> {
    root: PathRoot<'a, E>,
    members: Vec<&'a E::Selection>,
}

impl<'a, E: ExecutableDocument> Clone for Path<'a, E> {
    fn clone(&self) -> Self {
        Self {
            root: self.root,
            members: self.members.clone(),
        }
    }
}

impl<'a, E: ExecutableDocument> Path<'a, E> {
    pub fn new(root: PathRoot<'a, E>) -> Self {
        Self {
            root,
            members: Vec::new(),
        }
    }

    pub fn root(&self) -> &PathRoot<'a, E> {
        &self.root
    }

    pub fn with_selection(&self, selection: &'a E::Selection) -> Self {
        let mut clone = self.clone();
        clone.members.push(selection);
        clone
    }
}

pub enum PathRoot<'a, E: ExecutableDocument> {
    Operation(&'a E::OperationDefinition),
    Fragment(&'a E::FragmentDefinition),
}

impl<'a, E: ExecutableDocument> Clone for PathRoot<'a, E> {
    fn clone(&self) -> Self {
        match self {
            Self::Operation(o) => Self::Operation(o),
            Self::Fragment(f) => Self::Fragment(f),
        }
    }
}

impl<'a, E: ExecutableDocument> Copy for PathRoot<'a, E> {}

impl<'a, E: ExecutableDocument> Hash for PathRoot<'a, E> {
    fn hash<H: Hasher>(&self, state: &mut H) {
        match self {
            Self::Operation(o) => o.hash(state),
            Self::Fragment(f) => f.hash(state),
        }
    }
}

impl<'a, E: ExecutableDocument> PartialEq for PathRoot<'a, E> {
    fn eq(&self, other: &Self) -> bool {
        match (self, other) {
            (Self::Operation(l), Self::Operation(r)) if l == r => true,
            (Self::Fragment(l), Self::Fragment(r)) if l == r => true,
            _ => false,
        }
    }
}

impl<'a, E: ExecutableDocument> Eq for PathRoot<'a, E> {}

impl<'a, E: ExecutableDocument> Ord for PathRoot<'a, E> {
    fn cmp(&self, other: &Self) -> Ordering {
        match (self, other) {
            (Self::Fragment(l), Self::Fragment(r)) => l.cmp(r),
            (Self::Fragment(_), Self::Operation(_)) => Ordering::Greater,
            (Self::Operation(_), Self::Fragment(_)) => Ordering::Less,
            (Self::Operation(l), Self::Operation(r)) => l.cmp(r),
        }
    }
}

impl<'a, E: ExecutableDocument> PartialOrd for PathRoot<'a, E> {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}
