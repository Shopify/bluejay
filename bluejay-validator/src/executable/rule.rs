use crate::executable::Visitor;
use bluejay_core::definition::SchemaDefinition;
use bluejay_core::executable::ExecutableDocument;

pub trait Rule<'a, E: ExecutableDocument + 'a, S: SchemaDefinition + 'a>:
    Visitor<'a, E, S> + IntoIterator<Item = Self::Error>
{
    type Error;
}
