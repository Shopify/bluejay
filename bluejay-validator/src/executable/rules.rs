mod field_selection_merging;
mod field_selections;
mod leaf_field_selections;
mod lone_anonymous_operation;
mod named_operation_name_uniqueness;
mod operation_type_is_defined;
mod subscription_operation_single_root_field;

use crate::executable::{Error, Rule, Visitor};
use bluejay_core::definition::{SchemaDefinition, TypeDefinitionReferenceFromAbstract};
use bluejay_core::executable::{ExecutableDocument, OperationDefinitionFromExecutableDocument};
use field_selection_merging::FieldSelectionMerging;
use field_selections::FieldSelections;
use leaf_field_selections::LeafFieldSelections;
use lone_anonymous_operation::LoneAnonymousOperation;
use named_operation_name_uniqueness::NamedOperationNameUniqueness;
use operation_type_is_defined::OperationTypeIsDefined;
use paste::paste;
use std::iter::Chain;
use subscription_operation_single_root_field::SubscriptionOperationSingleRootField;

macro_rules! define_rules {
    ( $( $rule:ty ),* $(,)? ) => {
        paste! {
            pub struct Rules<'a, E: ExecutableDocument, S: SchemaDefinition> {
                $([<$rule:snake>]: $rule<'a, E, S>,)*
            }

            impl<'a, E: ExecutableDocument, S: SchemaDefinition> Rule<'a, E, S> for Rules<'a, E, S> {
                fn new(executable_document: &'a E, schema_definition: &'a S) -> Self {
                    Self {
                        $([<$rule:snake>]: $rule::new(executable_document, schema_definition),)*
                    }
                }
            }

            impl<'a, E: ExecutableDocument, S: SchemaDefinition> IntoIterator for Rules<'a, E, S> {
                type Item = Error<'a, E, S>;
                type IntoIter = chain_types!($(<$rule<'a, E, S> as IntoIterator>::IntoIter),*);

                fn into_iter(self) -> Self::IntoIter {
                    chain_iters!($(self.[<$rule:snake>].into_iter()),*)
                }
            }

            impl<'a, E: ExecutableDocument, S: SchemaDefinition> Visitor<'a, E, S> for Rules<'a, E, S> {
                fn visit_operation(&mut self, operation_definition: &'a OperationDefinitionFromExecutableDocument<E>) {
                    $(self.[<$rule:snake>].visit_operation(operation_definition);)*
                }

                fn visit_selection_set(&mut self, selection_set: &'a E::SelectionSet, r#type: &'a TypeDefinitionReferenceFromAbstract<S::TypeDefinitionReference>) {
                    $(self.[<$rule:snake>].visit_selection_set(selection_set, r#type);)*
                }

                fn visit_field(&mut self, field: &'a E::Field, r#type: &'a S::OutputTypeReference) {
                    $(self.[<$rule:snake>].visit_field(field, r#type);)*
                }
            }
        }
    };
}

macro_rules! chain_types {
    ( $first:ty, $( $rest:ty ),+ $(,)? ) => {
        Chain<chain_types!($($rest),+), $first>
    };
    ( $t:ty ) => { $t };
}

macro_rules! chain_iters {
    ( $first:expr, $( $rest:expr ),+ $(,)? ) => {
        chain_iters!($($rest),+).chain($first)
    };
    ( $iter:expr ) => { $iter };
}

define_rules!(
    NamedOperationNameUniqueness,
    LoneAnonymousOperation,
    SubscriptionOperationSingleRootField,
    FieldSelections,
    FieldSelectionMerging,
    OperationTypeIsDefined,
    LeafFieldSelections,
);
