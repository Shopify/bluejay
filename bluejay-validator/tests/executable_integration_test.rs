use bluejay_parser::{
    ast::definition::{DefinitionDocument, SchemaDefinition},
    ast::executable::ExecutableDocument,
    Error,
};
use bluejay_validator::executable::RulesValidator;

#[test]
fn test_error() {
    with_schema(|schema_definition| {
        insta::glob!("test_data/executable/error/*.graphql", |path| {
            let input = std::fs::read_to_string(path).unwrap();
            let executable_document =
                ExecutableDocument::parse(input.as_str()).expect("Document had parse errors");
            let errors = RulesValidator::validate(&executable_document, &schema_definition);
            let formatted_errors = Error::format_errors(input.as_str(), errors);
            insta::assert_snapshot!(formatted_errors);
        });
    });
}

#[test]
fn test_valid() {
    with_schema(|schema_definition| {
        insta::glob!("test_data/executable/valid/*.graphql", |path| {
            let input = std::fs::read_to_string(path).unwrap();
            let executable_document = ExecutableDocument::parse(input.as_str())
                .expect(format!("Document `{}` had parse errors", path.display()).as_str());
            let errors: Vec<_> = RulesValidator::validate(&executable_document, &schema_definition)
                .into_iter()
                .collect();
            assert!(
                errors.is_empty(),
                "Document `{}` had validation errors",
                path.display()
            )
        });
    });
}

fn with_schema<F: Fn(SchemaDefinition) -> ()>(f: F) {
    let s = std::fs::read_to_string("tests/test_data/executable/schema.graphql").unwrap();
    let definition_document =
        DefinitionDocument::parse(s.as_str()).expect("Schema had parse errors");
    let schema_definition =
        SchemaDefinition::try_from(&definition_document).expect("Schema had errors");
    f(schema_definition)
}
