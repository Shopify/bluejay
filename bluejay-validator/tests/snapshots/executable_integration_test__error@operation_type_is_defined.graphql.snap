---
source: bluejay-validator/tests/executable_integration_test.rs
expression: formatted_errors
input_file: bluejay-validator/tests/test_data/executable/error/operation_type_is_defined.graphql
---
Error: Schema does not define a mutation root
   ╭─[<unknown>:1:1]
   │
 1 │ mutation {
   │ ────┬───  
   │     ╰───── Schema does not define a mutation root
───╯

